### PR TITLE
Added version checker at the begining of execution

### LINKED ISSUE #74: Add "A New Release is Available" Check at the beginning of Job Execution
## Current State
We are currently shipping new releases regularly, and so a user coming back to cloud-concierge after even a week may be missing valuable updates.

## Desired State
During start-up, the container logs a colored message indicating if the user has the latest version of the container or should re-pull the image.

"You are currently running version v0.1.5 of cloud-concierge, [this is the latest version][the latest version is vXYZ, run docker pull dragondrop/cloud-concierge:latest to update.]"

### PR DESCRIPTION
**Pull request by AlexDragonDrop**
Resolves https://github.com/dragondrop-cloud/cloud-concierge/issues/74

### CODE REVIEW

**Comment by AlexDragonDrop** [comment]
Docker compose

Screenshot 2023-09-20 at 12 08 37

Screenshot 2023-09-20 at 12 13 47

Console binary execution

Screenshot 2023-09-20 at 12 14 31

Screenshot 2023-09-20 at 12 14 05

File: main/version_checker.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/fatih/color"
+	log "github.com/sirupsen/logrus"
+)
+
+const currentVersion = "v0.1.4"
```
Comment (GoodmanBen) [inline main/version_checker.go]:
```suggestion
const currentVersion = "v0.1.5"
```

File: main/version_checker.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	}
+}
+
+type GithubTagResponse struct {
+	TagName string `json:"tag_name"`
+}
+
+func (v *VersionChecker) getLatestCloudConciergeVersion() (string, error) {
+	log.Debugf("Checking for latest version of cloud-concierge...")
+	response, err := v.httpClient.Get("https://api.github.com/repos/dragondrop-cloud/cloud-concierge/releases/latest")
```
Comment (GoodmanBen) [inline main/version_checker.go]: Does this handle the following edge case:
- We push a new, pre-prod release (say v0.1.5-beta).

We basically want to guarantee that that the user is only recommended to upgrade to a production-ready release, not a beta release.

May need to remove the `/latest` suffix here and then iterate for the first tag that does not have a `-beta` suffix.

File: main/job.go [hunk 3]
```diff
@@ -24,7 +24,7 @@ import (
 
 // Job is an instance of a runnable dragondrop job.
 type Job struct {
-	// vcs is the implementation of interfaces.VCS for interacting with a remote version control system
+	// vcs is the implementation of interfaces.VCS for interacting with a remote latestStableVersion control system
```
Comment (GoodmanBen) [inline main/job.go]:
```suggestion
	// vcs is the implementation of interfaces.VCS for interacting with a remote version control system
```

File: main/job_config.go [hunk 4]
```diff
@@ -53,7 +53,7 @@ type JobConfig struct {
 	// to the right state files.
 	NLPEndpoint string `default:"https://us-east4-dragondrop-prod.cloudfunctions.net/nlpengine-endpoint-prod"`
 
-	// TerraformVersion is the version of Terraform used.
+	// TerraformVersion is the latestStableVersion of Terraform used.
```
Comment (GoodmanBen) [inline main/job_config.go]: "Control R" replace all error 😄

File: main/job_config.go [hunk 5]
```diff
@@ -68,7 +68,7 @@ type JobConfig struct {
 	// WorkspaceDirectories is a slice of directories that contains terraform workspaces within the user repo.
 	WorkspaceDirectories terraformWorkspace.WorkspaceDirectoriesDecoder `required:"true"`
 
-	// Provider is a map between a cloud provider and the version for that provider.
+	// Provider is a map between a cloud provider and the latestStableVersion for that provider.
```
Comment (GoodmanBen) [inline main/job_config.go]: "Control R" replace all error 😄

File: main/job_config_inferred.go [hunk 6]
```diff
@@ -22,7 +22,7 @@ type InferredData struct {
 	// Provider is the name of the cloud provider (aws, azurerm, google, etc.).
 	Provider terraformValueObjects.Provider `required:"true"`
 
-	// VCSSystem is the name of the version control system (github, gitlab, bitbucket, etc.).
+	// VCSSystem is the name of the latestStableVersion control system (github, gitlab, bitbucket, etc.).
```
Comment (GoodmanBen) [inline main/job_config_inferred.go]: "Control R" replace all error 😄

File: main/job_config_inferred.go [hunk 7]
```diff
@@ -33,7 +33,7 @@ type InferredData struct {
 func getInferredData(config JobConfig) (InferredData, error) {
 	provider, err := getProviderFromProviderVersion(config.Provider)
 	if err != nil {
-		return InferredData{}, fmt.Errorf("[error getting the provider value from provider version]%w", err)
+		return InferredData{}, fmt.Errorf("[error getting the provider value from provider latestStableVersion]%w", err)
```
Comment (GoodmanBen) [inline main/job_config_inferred.go]: "Control R" replace all error 😄

File: main/job_config_inferred.go [hunk 8]
```diff
@@ -199,7 +199,7 @@ func getGoogleCredential(jobID string) (terraformValueObjects.Credential, error)
 	return terraformValueObjects.Credential(""), nil
 }
 
-// getProviderFromProviderVersion determines the provider from the input provider version
+// getProviderFromProviderVersion determines the provider from the input provider latestStableVersion
```
Comment (GoodmanBen) [inline main/job_config_inferred.go]: "Control R" replace all error 😄

File: main/version_checker.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+const currentVersion = "v0.1.5"
+const footer = "###########################################################################################################################################################"
+
+type VersionChecker struct {
+	// httpClient is a http client shared across all http requests within this package.
+	httpClient http.Client
+}
+
+func (v *VersionChecker) checkCloudConciergeVersion() {
```
Comment (GoodmanBen) [inline main/version_checker.go:24]: Missing function comment

File: main/version_checker.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+		color.Green(fmt.Sprintf("############################### %s ##################################", message))
+		color.Green(footer)
+	}
+}
+
+type GithubRelease struct {
+	TagName string `json:"tag_name"`
+}
+
+func (v *VersionChecker) getLatestStableCloudConciergeVersion() (string, error) {
```
Comment (GoodmanBen) [inline main/version_checker.go:52]: Missing function comment

File: main/version_checker.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+	for _, release := range githubReleases {
+		if !strings.Contains(release.TagName, "-beta") {
+			return release.TagName, nil
+		}
+	}
+
+	return "", nil
+}
+
+func (v *VersionChecker) getCloudConciergeVersionMessage(currentVersion, latestVersion string) (string, string) {
```
Comment (GoodmanBen) [inline main/version_checker.go:85]: Missing function comment

**Review by GoodmanBen — COMMENTED** [review]
A few things to clean up then looks good to me.

**Review by GoodmanBen — APPROVED** [review]

**Pull request merged**